dali/operators/segmentation/random_object_bbox.cc

```diff
@@ -0,0 +1,442 @@
+    case RandomObjectBBox::Out_AnchorShape:
+      for (int i = 0; i < ndim; i++) {
+        out1.data[sample_idx][i] = start[i];
+        out2.data[sample_idx][i] = end[i] - start[i];
+      }
+      break;
+    case RandomObjectBBox::Out_StartEnd:
+      for (int i = 0; i < ndim; i++) {
+        out1.data[sample_idx][i] = start[i];
+        out2.data[sample_idx][i] = end[i];
+      }
+      break;
+    default:
+      assert(!"Unreachable code");
+  }
+}
+
+template <typename Box>
+void StoreBox(const OutListCPU<int, 1> &out1,
+              const OutListCPU<int, 1> &out2,
+              RandomObjectBBox::OutputFormat format,
+              int sample_idx, Box &&box) {
+  StoreBox(out1, out2, format, sample_idx, box.lo, box.hi);
+}
+
+void RandomObjectBBox::GetClassesAndWeightsArgs(
+      ClassVec &classes, WeightVec &weights, int &background, int sample_idx) {
+  background = background_[sample_idx].data[0];
+  if (ignore_class_)
+    return;  // we don't care about classes at all
+
+  if (classes_.IsDefined()) {
+    const auto &cls_tv = classes_[sample_idx];
+    int ncls = cls_tv.shape[0];
+    classes.resize(ncls);
+    if (!weights_.IsDefined()) {
+      weights.clear();
+      weights.resize(ncls, 1.0f);
+    }
+
+    int min_class_label = ncls ? cls_tv.data[0] : 0;
+    bool choose_different_background = false;
+    for (int i = 0; i < ncls; i++) {
+      classes[i] = cls_tv.data[i];
+      if (classes[i] < min_class_label)
+        min_class_label = classes[i];
+
+      if (classes[i] == background) {
+        if (background_.IsDefined()) {
+          // Background was explicitly specified this way - that's an error
+          DALI_FAIL(make_string("Class label ", classes[i],
+            " coincides with background label - please specify a different background label or "
+            " remove it from your list of foreground classes."));
+        } else {
+          // Not specified? Pick a different one.
+          choose_different_background = true;
+        }
+      }
+    }
+    if (choose_different_background) {
+      // This will yield incorrect result if the set of classes contains both INT_MIN and INT_MAX,
+      // which is a sacrifice I am willing to make.
+      background = min_class_label - 1;
+    }
+  }
+  if (weights_.IsDefined()) {
+    const auto &cls_tv = weights_[sample_idx];
+    int ncls = cls_tv.shape[0];
+    weights.resize(ncls);
+    for (int i = 0; i < ncls; i++)
+      weights[i] = cls_tv.data[i];
+    if (!classes_.IsDefined()) {
+      classes.resize(ncls);
+      int cls = 0;
+      for (int i = 0; i < ncls; i++, cls++) {
+        if (cls == background)
+          cls++;
+        classes[i] = cls;
+      }
+    }
+  }
+}
+
+template <int ndim>
+int RandomObjectBBox::PickBox(span<Box<ndim, int>> boxes, int sample_idx) {
+  auto beg = boxes.begin();
+  auto end = boxes.end();
+  if (threshold_.IsDefined()) {
+    vec<ndim, int> threshold;
+    const int *thresh = threshold_[sample_idx].data;
```

Comment:
```
auto threshold = *reinterpret_cast<const vec<ndim, int>*>(threshold_[sample_idx].data);
```
This would save you the loop in line 273

Reply:
It's technically undefined behavior to type-pun the pointer this way.

Reply:
Fair enough. I was pointing that out because we've used that in the past